server/src/main/java/org/opensearch/search/query/QueryPhase.java

```diff
@@ -309,6 +305,29 @@ static boolean executeInternal(SearchContext searchContext, QueryPhaseSearcher q
         }
     }
 
+    /**
+     * Create runnable which throws {@link TimeExceededException} when the runnable is called after timeout + runnable creation time
+     * exceeds currentTime
+     * @param relativeMsTimeSupplier supplier of currentTime in milliseconds which is okay to be cached
+     * @param timeout the max time duration in milliseconds before throwing {@link TimeExceededException}
+     * @return the created runnable
+     */
+    static Runnable createQueryTimeoutChecker(LongSupplier relativeMsTimeSupplier, final long timeout) {
+        /* for startTime, relative non-cached time must be used to prevent false positive timeouts.
+        * Using cached time for startTime will fail and produce false positive timeouts when maxTime = (startTime + timeout) falls in
+        * next time cache slot(s) AND time caching lifespan > passed timeout */
+        final long startTime = TimeValue.nsecToMSec(System.nanoTime());
```

Comment:
@aabukhalil I think we are focusing only on one aspect of the problem rather than whole:
 - if `thread_pool.estimated_time_interval` **is** set to `0`, no changes needed - the original implementation will work right away
 - if `thread_pool.estimated_time_interval` **is not** set to `0`, we may have an issues here but only if the `timeout` value is significantly smaller than `thread_pool.estimated_time_interval`

Instead of changing `QueryPhase`, I would like to suggest introducing new method `getRelativeTimeInMillisForTimeout(final long timeout)` (or even modifying the existing method `getRelativeTimeInMillis` to accept `timeout` argument, seems like it is used only in `QueryPhase`). Inside, depending on the `thread_pool.estimated_time_interval` and `timeout` value the decision could be made to either use cached time provider or fallback to `System.nanoTime()`. WDYT?

Reply:
@reta thank you for reviewing this and providing feedback :) 
Can you please elaborate on your concern on the reason of introducing more branching logic to compute initial time (which depend on cache lifespan and timeout value) as you suggested instead of using one logic which will always work with no branching ?

And yeah I have concern of calling `System.nanoTime` directly from `QueryPhase` but I think adding additional method `getRelativeTimeInMillisForTimeout` or override to `getRelativeTimeInMillis(long timeout)` will be more confusing to which one to use. (It will be harder to decide which method/override to use  in [here](https://github.com/aabukhalil/OpenSearch/blob/f604902143505da066dfbe8f660f3d3914f148ab/server/src/main/java/org/opensearch/search/query/QueryPhase.java#L324) for example or any caller and eventually need to look at implementation details to decide which one). My suggestion is to introduce method like `getNonCachedRelativeTimeInMillis` or override like `getRelativeTimeInMillis(boolean useCache)` where it is clear for caller what is the difference between them and let caller decide on branching logic. What do you think ?

Reply:
@reta yeah I agree that time provider should be the only responsible module for time but I'm asking about  your concern of adding new branch (if timeout < cache lifespan then return non cached time else fallback to default implementation) instead of always returning non cached time and without promising that we have actual control of time precision because I think we have cached time and non cached time actually. WDYT ?

Reply:
@aabukhalil sorry missed that, I don't know the full story behind caching, but may be we could measure the overhead of calling `System.nanoTime` vs query execution time? (like [1] fe) Calling `System.nanoTime` directly and dropping all the caching impl would make things simpler for sure.

[1] https://bugs.openjdk.java.net/browse/JDK-6440430

Reply:
> I don't know the full story behind caching, but may be we could measure the overhead of calling System.nanoTime vs query execution time?

I don't know the history either, but the pattern of caching the time seems to have been introduced [over 10 years ago](https://github.com/opensearch-project/OpenSearch/commit/a1796c3408d1847e44a7d55ac7c048621aff4041). It is very possible that the world has changed since then and the caching behavior is no longer needed. I'd love to remove the caching to make things simpler, but it might be hard to have confidence that doing so won't introduce a regression on some OS/hardware combination even if it works great on a modern linux setup. @reta, what sort of testing would give you confidence that we could just remove the caching all together?

Reply:
@reta `System.nanoTime` is known to be a slow operation and that's why it is cached. But the caching advantage is still applicable here because [getting time here](https://github.com/aabukhalil/OpenSearch/blob/f604902143505da066dfbe8f660f3d3914f148ab/server/src/main/java/org/opensearch/search/query/QueryPhase.java#L324) is still using the cached time and which will be called almost per matched document. So the overhead I'm adding is one `System.nanoSecond` per query when timeout is set and after merging this I need to look on the performance impact of this single call. For now I will refactor this to delegate calling `System.nanoSecond` to SearchContext.

Reply:
> @reta, what sort of testing would give you confidence that we could just remove the caching all together?

@aabukhalil @andrross  there is this excellent analysis [1] from OpenJDK developer(s) regarding the `System.nanoTime()` usage, the takeaway (one of) is: "nanoTime is not dirt cheap; at best, you can hope for 15-30 ns per call" (is highly depends on operating system as well). 

[1] https://shipilev.net/blog/2014/nanotrusting-nanotime/

Reply:
@reta @andrross Yes `nanoTime` is not cheap and seems nothing has changed since these analysis were made. According to a micro benchmark I've just made (will include it's code in next revision), calling `System.nanoTime` has latency of ~30 nanoseconds while accessing volatile long variable and incrementing it has latency of ~3 nanoseconds (which is upper bound of what is happening in [ThreadPool time cache](https://github.com/opensearch-project/OpenSearch/blob/1ceff286a0ead3164c99b66332d4e008218bb338/server/src/main/java/org/opensearch/threadpool/ThreadPool.java#L595) ). 

The benchmark was run on Linux on EC2 c6a.12xlarge (AMD processor), Linux on EC2 m5.12xlarge (Intel processor), Macbook locally (Intel(R) Core(TM) i7-9750H) and Windows locally (Intel(R) Core(TM) 7-10750H) and all shows the consistent behavior of 30ns latency when calling `nanoTime`.

My call here is that the advantage of cached time optimization is still applicable as [getting time here](https://github.com/aabukhalil/OpenSearch/blob/f604902143505da066dfbe8f660f3d3914f148ab/server/src/main/java/org/opensearch/search/query/QueryPhase.java#L324) is still using the cached time which will be called almost per matched document in query result which even the 30ns latency MIGHT be amplified depending on match set size because of how [checkCancelled](https://github.com/opensearch-project/OpenSearch/blob/1ceff286a0ead3164c99b66332d4e008218bb338/server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java#L419) is called as currently there are 21 caller of that method and some of them with no sampling (e.g. 100K matching documents might add up to 30 millisecond to query's `took` when timeout is set with no time caching). 

My change is called once only per query where timeout is set which will add ~30ns to overall search latency per query which I don't think will impact overall system latency or throughput and still I need to follow up on performance test after merging this to confirm that there is no impact made. What do you think ?